llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp

```diff
@@ -1155,12 +1155,16 @@ unsigned getMinNumVGPRs(const MCSubtargetInfo *STI, unsigned WavesPerEU) {
   return std::min(MinNumVGPRs, AddrsableNumVGPRs);
 }
 
-unsigned getMaxNumVGPRs(const MCSubtargetInfo *STI, unsigned WavesPerEU) {
+unsigned getMaxNumVGPRs(const MCSubtargetInfo *STI, unsigned WavesPerEU,
+                        bool WholeRegisterFile) {
   assert(WavesPerEU != 0);
 
-  unsigned MaxNumVGPRs = alignDown(getTotalNumVGPRs(STI) / WavesPerEU,
-                                   getVGPRAllocGranule(STI));
-  unsigned AddressableNumVGPRs = getAddressableNumVGPRs(STI);
+  unsigned MaxNumVGPRs =
+      alignDown(getTotalNumVGPRs(STI, WholeRegisterFile) / WavesPerEU,
```

Comment:
This looks wrong to me. The computation shall just return a number of registers allowed for occupancy, it shall not be halved. Then it is clamped below to account for the WholeRegisterFile.

Reply:
I see that getTotalNumVGPRs should always return the total number of VGPRs.

But, I think we should still adjust the halved value against the number of waves? Otherwise we may end up in our current situation where we say wavesPerEU of 3 is allowed 168 arch vgprs and 168 acc vgrps (this is actually occupancy 1). The latest version pre-clamps getTotalNumVGPRs to getTotalNumArchVGPRs when WholeRegisterFile is not requested.

Reply:
This assumes equal allocation. I.e. assume you want occupancy 2 and you have no agprs at all. You will end up with the limit of 128 vgprs, but you can use 256.

Reply:
Ah, yes I see thanks. I will account for this.